applications/finetune-quora-embeddings/embed.py

```diff
@@ -0,0 +1,284 @@
+    from datasets import load_dataset
+
+    dataset_path = f"{DATASET_DIR}/{DATASET_NAME}"
+
+    if os.path.exists(dataset_path):
+        return
+
+    dataset = load_dataset(DATASET_NAME)
+
+    dataset.save_to_disk(dataset_path)
+    DATASET_VOLUME.commit()
+
+
+@stub.function(
+    image=image,
+    volumes={DATASET_DIR: DATASET_VOLUME},
+    secrets=[
+        Secret.from_name("openai"),
+        Secret.from_name("cohere"),
+    ],
+    timeout=86400,
+)
+async def embed_dataset(model_name: str):
+    from datasets import load_from_disk, concatenate_datasets
+    import math
+    import time
+
+    # We verify if the model has already been embedded
+    if determine_if_model_has_generated_embeddings(model_name):
+        print(f"Embedding has already been generated for {model_name}")
+        return
+
+    combined_num_rows = 408651  # Extract
+    # First Load the model
+    if model_name in MODELS:
+        model: model_types = "HuggingFace"
+        embed_model = EmbeddingModel.from_hf(
+            model_name, math.ceil(combined_num_rows / BATCH_SIZE_CONFIG["HuggingFace"])
+        )
+    elif model_name in COHERE_MODELS:
+        embed_model = EmbeddingModel.from_cohere(model_name)
+        model = "Cohere"
+    elif model_name in OPENAI_MODELS:
+        model: model_types = "OpenAI"
+        embed_model = EmbeddingModel.from_openai(model_name, SEMAPHORE_LIMITS["OpenAI"])
+    else:
+        raise ValueError(
+            f"Invalid Model of {model_name} was supplied to embed_dataset function"
+        )
+
+    dataset = load_from_disk(f"{DATASET_DIR}/{DATASET_NAME}")
+    test_dataset = dataset["test"]
+    train_dataset = dataset["train"]
+    combined_dataset = concatenate_datasets([test_dataset, train_dataset])
+
+    sentence_to_id_map = dict()
+
+    for idx in range(3):
+        print(f"Iteration {idx}")
+        try:
+            sentences = get_unique_sentences(
+                combined_dataset, sentence_to_id_map, BATCH_SIZE_CONFIG[model]
+            )
+            sentence_embeddings = await embed_model.embed(sentences)
+            if len(sentence_embeddings) == combined_num_rows:
+                break
+        except Exception as e:
+            print(
+                f"Encountered Exception of {e}. Retrying embedding eneration in 30 seconds again"
+            )
+            time.sleep(30)
+
+    return update_dataset_with_embeddings(
+        train_dataset,
+        sentence_to_id_map,
+        sentence_embeddings,
+    ), update_dataset_with_embeddings(
+        test_dataset,
+        sentence_to_id_map,
+        sentence_embeddings,
+    )
+
+
+@stub.function(image=image, volumes={DATASET_DIR: DATASET_VOLUME}, timeout=2400)
+def generate_embeddings():
+    import pyarrow as pa
+    import os
+
+    test = OPENAI_MODELS[1:]
+
+    if not os.path.exists(CACHE_DIRECTORY):
+        os.makedirs(CACHE_DIRECTORY)
+
+    for model_name, res in zip(test, embed_dataset.map(test, order_outputs=True)):
```

Comment:
no need to zip if embed_dataset returns `test` as a tuple. then order_outputs can be false and we can return sooner